### PR TITLE
Implement full topological sort

### PR DESCRIPTION
**Pull request by netalondon**
@DavidSouther turns out we did need a proper topological sort.
This is part of #327

### CODE REVIEW

**Review by DavidSouther — APPROVED** [review]

**Review by DavidSouther — CHANGES_REQUESTED** [review]
(Clicked the wrong button)

File: simulator/src/chip/chip.test.ts [hunk 1]
```diff
@@ -670,8 +671,43 @@ describe("Chip", () => {
...
+        {
+          from: { name: "a", start: 0, width: 16 },
+          to: { name: "out", start: 0, width: 16 },
+        },
+      ]);
+    }
+  }
+  const fooC = new FooC();
+  fooC.sortParts();
+  expect(fooC.parts).toEqual([fooC.register, fooC.inc16]);
```
Comment (DavidSouther) [inline simulator/src/chip/chip.test.ts:712]: This currently passes on main without calling sortParts, so I don't think it shows the bug from #327

File: simulator/src/chip/builder.ts [hunk 2]
```diff
@@ -220,6 +211,11 @@ class ChipBuilder {
     if (isErr(result)) {
       return result;
     }
+
+    this.chip.sortParts();
+    for (const part of this.chip.parts) {
+      part.subscribeToClock();
```
Comment (DavidSouther) [inline simulator/src/chip/builder.ts:217]: This PR has gotten big enough it's trying to do two things - can we separate out the subscription updates from the sort updates?

File: simulator/src/chip/graph.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  sort(start: number) {
+    const visited = new Set<number>();
+    const visiting = new Set<number>();
+    const sorted: number[] = [];
+    const stack: NodeState[] = [{ node: start, isReturning: false }];
+
+    while (stack.length > 0) {
+      const { node, isReturning } = stack.pop()!;
```
Comment (DavidSouther) [inline simulator/src/chip/graph.ts:23]: assertExists from `@davidsouther/jiffies` is a low overhead tool to not need the non-null assertion.

File: simulator/src/chip/graph.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        sorted.push(node);
+      } else if (!visited.has(node)) {
+        // If we are visiting this node for the first time
+        if (visiting.has(node)) {
+          continue;
+        }
+        visiting.add(node);
+
+        // Re-push this node to handle it on return
+        stack.push({ node, isReturning: true });
```
Comment (DavidSouther) [inline simulator/src/chip/graph.ts:37]: Oh very elegant way to handle this

**Review by DavidSouther — CHANGES_REQUESTED** [review]

**Review by DavidSouther — COMMENTED** [review]
Let me know when this is rebased and the minimal test is in place.

**Comment by netalondon** [comment]
> Let me know when this is rebased and the minimal test is in place.

I updated the test and now I believe it fails on main.
Did I miss something?

File: simulator/src/chip/chip.test.ts [hunk 5]
```diff
@@ -641,7 +641,10 @@ describe("Chip", () => {
     }
 
     const fooA = new FooA();
-    expect(fooA.parts).toEqual([fooA.notB, fooA.notA]);
```
[the remark this answers is not in the document]
Comment (DavidSouther) [reply simulator/src/chip/chip.test.ts:644]: I updated the test in my branch to just look at the IDs, which is the important piece.

**Comment by DavidSouther** [comment]
I opened #339 to keep the insertion sort, which I maintain is the neater and cleaner approach. But this PR was invaluable in tracking down a test case to make sure we got it right both here and for #327

**Pull request closed without merging**